### PR TITLE
feat: implement parser for use in test harness

### PR DESCRIPTION
**Pull request by bitwalker**
**NOTE:** This PR builds on the work done in #43, but with some refactoring to address various nits and bugs that were outstanding. With the changes made so far, the parser is working as expected for now, but this is being opened as a draft, as there are a few small issues/todos remaining:

- [x] Need to ensure that identifiers which conflict with keywords in the IR syntax are quoted during formatting. Parsing quoted identifiers is now implemented, so we just need to make sure the other direction is covered.
- [x] Need to implement conversion from the parser AST to the actual IR representation, where we'll do all of the necessary validation of the parsed IR. This should be done before merging or we can't do much of anything useful with it.
- [x] During conversion, we need to convert immediates to the correct `Immediate` variant. If we can't do that with the current syntax, we'll add type suffixes to immediates so that we can always roundtrip immediates through the text format.
- [ ] Need to add support to the parser for data segments. This can be done in a later PR, but will be needed sooner rather than later.
- [ ] Need to add support to the parser for struct attributes (i.e. packing, alignment). This can be done in a later PR.
- [x] Need to synchronize some outstanding small tweaks to the formatter based on what we require in the parser

I've also imposed a restriction on the textual format in the near term, which is that one may not use arbitrary identifiers for values/blocks - you are required to use the same format as used when emitting the textual format (i.e. `vN` for values, where `N` is a number; `blkN` for blocks, and so on).

While some small tweaks have been made to the format, most things have largely been preserved. The main difference is the introduction of line terminators for instructions, i.e. all instructions must be followed by `;`, except `switch` which is wrapped in `{`/`}`.

### CODE REVIEW

**Comment by bitwalker** [comment]
@greenhat This is ready for review now. I'm expecting a couple of tests to fail due to changes in the formatter, so I'll fix those before we merge this, but the rest can be reviewed in isolation. I'm not super stoked about how messy the parser code is right now, but I don't really have time to spend on improving it at the moment, so I think I'll plan to revisit that in the future.

**Review by greenhat — APPROVED** [review]
Looking good! I did not dive deep into the parser internals since it would take a significant amount of time. I was more focused on how it affects other parts. Please check my comments.

File: hir/src/globals.rs [hunk 1]
```diff
@@ -440,11 +461,17 @@ impl GlobalVariableTable {
     ///
     /// It is expected that the caller has already guaranteed that the name of the given global variable
     /// is not present in the table, and that all validation rules for global variables have been enforced.
-    unsafe fn insert(&mut self, mut data: GlobalVariableData) -> GlobalVariable {
+    pub(crate) unsafe fn insert(&mut self, mut data: GlobalVariableData) -> GlobalVariable {
         let name = data.name;
         // Allocate the data in the arena
-        let gv = self.arena.alloc_key();
-        data.id = gv;
+        dbg!(data.id, GlobalVariable::default());
```
[the remark this answers is not in the document]
Comment (bitwalker) [reply hir/src/globals.rs:467]: Ah good catch!

File: hir/src/parser/ast/mod.rs [hunk 2]
```diff
@@ -0,0 +1,1125 @@
...
+
+    // Translate instruction data from AST representation to HIR
+    let data = match inst.ty {
+        InstType::BinaryOp {
+            opcode: op,
+            overflow,
+            operands: [Operand::Value(lhs), Operand::Value(rhs)],
+        } => Some(Instruction::BinaryOp(BinaryOp {
+            op,
+            overflow,
```
[the remark this answers is not in the document]
Comment (bitwalker) [reply hir/src/parser/ast/mod.rs:428]: Most binary operators do not have overflow options, and of those that do, some of them only a subset of options (e.g. `checked` vs `unchecked`). The overflow variant was not being emitted in the text format previously though, so we were just ignoring it. Now however, it is being emitted, and we only want to display overflow for instructions where it applies. On the flip side, we could create new instruction variants for instructions with overflow, but it seems better to make it an optional field of the existing variants.

I realized this morning I need to update a few builders to ensure that `overflow` is always set for instructions which expect to have it, so we either need to decide what the default is, or remove builders like `add` which don't specify any overflow.

As for defaulting to `Checked` or `Unchecked`, I think in general the expected behavior would be `Checked` (for example, Rust's `+` asserts on overflow). `Wrapping` is another alternative, and I think that's how Wasm behaves. Unchecked operations are unsafe, so should be opt-in, IMO.

**Comment by bitwalker** [comment]
@greenhat I've made some adjustments to the builder to be more precise about which binary/unary ops support which types of overflow, and updated all uses accordingly. You may want to check my changes to `code_translator.rs`, but I believe those choices match the Wasm semantics.

I also discovered that the handling of the `Overflowing` flag was not correct, and have addressed that in a few places. There are a few other small adjustments/fixes, but those are probably not interesting.

I'm going to merge this as-is, but just wanted to make you aware of those changes so you can review them if you like, just open an issue (or a PR) if you find anything that needs to be changed.

**Pull request merged**